tests/test_workers.py

```diff
@@ -257,3 +291,52 @@ async def test_aclose_pending_outbound(nursery, worker, connector_factory):
     with fail_after(CLOSED_WAIT_TIME):
         await worker.aclose()
     assert worker.is_closed
+
+
+@pytest.mark.worker_class.with_args(SlowSetupWorker)
+async def test_connector_setup_race(nursery, worker, connector_factory):
+    """
+    All connectors must be set up and available to publish before any messages
+    are consumed.
+    """
+    # In order to send a message that the worker will receive at startup, we
+    # need to create and bind the relevant queue. The easiest way to do that is
+    # to create a a matching connector. We then close that connector so it
+    # doesn't consume the message before the worker starts.
+    ri_ri = await connector_factory.setup_ri("ri")
+    await ri_ri.conn.aclose_consumers()
+
+    ro_ri = await connector_factory.setup_ro("ri")
+    ri_ro = await connector_factory.setup_ri("ro")
+
+    # Publish a message before we start the worker so that the first connector
+    # receives it as soon as it's able to receive.
+    await ro_ri.publish_inbound(mkmsg("hi"))
+    await sleep(0.1)
+
+    await worker.setup()
+
+    # If all went well, we get no exception and we receive the message through
+    # the connector that was only set up later.
+    assert (await worker.exc.receive()) is None
+    assert (await ri_ro.consume_inbound()).content == "hi"
+
+
+@pytest.mark.worker_class.with_args(SlowSetupWorker)
+async def test_connector_setup_call_start_twice(connector_factory):
```

Comment:
We're not setting up a worker for this test, so we can drop the decorator. (Same for the next test.)